localstack/aws/connect.py

```diff
@@ -0,0 +1,473 @@
+        without affecting the caching mechanism.
+        """
+        return client
+
+    # TODO @cache here might result in a memory leak, as it keeps a reference to `self`
+    # We might need an alternative caching decorator with a weak ref to `self`
+    # Otherwise factories might never be garbage collected
+    @cache
+    def _get_client(
+        self,
+        service_name: str,
+        region_name: str,
+        use_ssl: bool,
+        verify: bool,
+        endpoint_url: str,
+        aws_access_key_id: str,
+        aws_secret_access_key: str,
+        aws_session_token: str,
+        config: Config,
+    ) -> BaseClient:
+        """
+        Returns a boto3 client with the given configuration, and the hooks added by `_get_client_post_hook`.
+        This is a cached call, so modifications to the used client will affect others.
+        Please use another instance of the factory, should you want to modify clients.
+        Client creation is behind a lock as it is not generally thread safe.
+
+        :param service_name: Service to build the client for, eg. `s3`
+        :param region_name: Name of the AWS region to be associated with the client
+            If set to None, loads from botocore session.
+        :param aws_access_key_id: Access key to use for the client.
+            If set to None, loads from botocore session.
+        :param aws_secret_access_key: Secret key to use for the client.
+            If set to None, loads from botocore session.
+        :param aws_session_token: Session token to use for the client.
+            Not being used if not set.
+        :param endpoint_url: Full endpoint URL to be used by the client.
+            Defaults to appropriate LocalStack endpoint.
+        :param config: Boto config for advanced use.
+        :return: Boto3 client.
+        """
+        with self._create_client_lock:
+            client = self._session.client(
+                service_name=service_name,
+                region_name=region_name,
+                use_ssl=use_ssl,
+                verify=verify,
+                endpoint_url=endpoint_url,
+                aws_access_key_id=aws_access_key_id,
+                aws_secret_access_key=aws_secret_access_key,
+                aws_session_token=aws_session_token,
+                config=config,
+            )
+
+        return self._get_client_post_hook(client)
+
+    #
+    # Boto session utilities
+    #
+    def _get_session_region(self) -> str:
+        """
+        Return AWS region as set in the Boto session.
+        """
+        return self._session.region_name
+
+    def _get_region(self) -> str:
+        """
+        Return the AWS region name from following sources, in order of availability.
+        - LocalStack request context
+        - LocalStack default region
+        - Boto session
+        """
+        return (
+            get_region_from_request_context()
+            or config.DEFAULT_REGION
```

Comment:
Is it really necessary to fall back to the default region here? What would be a use-case for that?
If we can avoid the default region, the `_get_session_region` would also make sense again (as it doesn't right now, see the TODO below).

Reply:
I will inverse the ranking for now, the last fallback will be the config variable. I guess it makes sense, if no other information is available to have a fallback but usually I would encourage explicitly setting the region anyway.

In cases where no request context is available (for example threads polling something) it might be interesting to have this fallback, but especially in those cases, a region should be set manually.